### PR TITLE
[PFT-222, PFT-224] Newly designed Periflow Python SDK

### PR DESCRIPTION
**Pull request by DifferentSC**
This PR includes

* Newly designed SDK API (Very Simple!)
* Unit test for newly designed API
* Updated CIFAR-10 example (transformer example is currently broken)

The newly designed APIs are very simple like...

```python
from periflow_sdk import periflow as pf

pf.init(...)
for data in dataloader:
  with pf.train_step():
    ...
    pf.save(...)
```

### CODE REVIEW

File: src/periflow_sdk/__init__.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+from threading import Thread
+from typing import Callable, Dict, List
+import logging
+from pathlib import Path
+import json
+from contextlib import contextmanager
+
+import torch
+from periflow_sdk.comm.ipc import IpcCommPurpose, CommResultStatus, get_default_ipc_channel
+from periflow_sdk.comm.errors import IpcTimeoutException, IpcConnectionFailureException
```
Comment (ktaebum) [inline src/periflow_sdk/__init__.py:19]: import order?
```
import atexit
import copy
import json
import logging
import os
import sys
import time
from contextlib import contextmanager
from enum import Enum
from pathlib import Path
from threading import Thread
from typing import Callable, Dict, List

import torch

from periflow_sdk.comm.errors import IpcTimeoutException
from periflow_sdk.comm.ipc import IpcCommPurpose, CommResultStatus, get_default_ipc_channel
```

File: src/periflow_sdk/__init__.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    EMERGENCY = "EMERGENCY"
+
+
+class TrainingManager:
+    """ The training wrapper class for general PyTorch training code.
+    Notes: This
+    """
+    def __init__(self,
+                 log_file_name: str = None,
+                 is_local: bool = None,
```
[the remark this answers is not in the document]
Comment (ktaebum) [reply src/periflow_sdk/__init__.py:35]: Do we have to get these as input argumetns? Because a user does not call the constructor explicitly

File: src/periflow_sdk/__init__.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                 teardown_at_exit: bool = True):
```
Comment (ktaebum) [inline src/periflow_sdk/__init__.py:36]: What is the purpose when `teardown_at_exit` is `False`?

File: src/periflow_sdk/__init__.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+                    raise RuntimeError(f"Invalid IPC message from FTModule: {ack}")
+
+                # If emergency save is done, terminate the training process.
+                if self._is_saved and self._save_method is SaveType.EMERGENCY:
+                    sys.exit()
+
+            except IpcConnectionFailureException as ipc_connection_failure:
+                raise RuntimeError("IPC connection between training manager and FTModule is broken.") \
+                    from ipc_connection_failure
+        self._is_step_started = False
```
Comment (ktaebum) [inline src/periflow_sdk/__init__.py:163]: If we set `_step_start_time` as `None` when `end_step` is called, we can make `_is_step_started` as
```
@property
def _is_step_started(self) -> bool:
    return self._step_start_time is not None
```

File: src/periflow_sdk/__init__.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        if "step" not in new_msg:
+            new_msg["step"] = self._cur_step
+        if self._is_local:
+            self._local_log(new_msg)
+        else:
+            self._metric_ipc_channel.write(new_msg)
+
+    def save(self, obj, path: str, async_save: bool = False):
+        assert not self._is_saved, "You cannot call `pf.save()` twice within a training step."
+        assert self._is_step_started, "You can only call `pf.save()` within a training step scope."
```
Comment (ktaebum) [inline src/periflow_sdk/__init__.py:193]: Do we have to enforce saving the checkpoint within a training step?
When users want to use the JIT compilation of PyTorch (torch.trace) or TensorFlow (tf.function),
many people do not include saving checkpoint in the training step function (due to correct compilation).
Furthermore, there are some cases that
- check validation loss
- if validation loss is the lowest, save checkpoint

In that case, our train_step would include the time of validation.
How about removing this constraint, and moving the communication of the `end_step` function into the `save` function?

**Review by ktaebum — COMMENTED** [review]
Thanks for your work 🙂 . I've just left my first review.

File: src/periflow_sdk/__init__.py (hunk 5 above)
Comment (DifferentSC) [reply src/periflow_sdk/__init__.py:193]: We've decided to leave this for now.

File: src/periflow_sdk/__init__.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    def is_emergency_save(self):
+        return self._emergency_save_step == self._cur_step
+
+    def _local_log(self, msg):
+        mode = "a" if self._has_locally_logged else "w"
+        with self._log_path.open(mode=mode) as log_file:
+            log_file.write(f"{json.dumps(msg)}\n")
+        self._has_locally_logged = True
+
+    def log(self, msg: Dict):
```
Comment (Algy) [inline src/periflow_sdk/__init__.py:182]: The name seems ambiguous. log -> metric?

File: src/periflow_sdk/__init__.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        try:
+            msg = self._emergency_save_ipc_channel.read(timeout=None)
+        except (IpcTimeoutException, IpcConnectionFailureException):
+            pass
+        else:
+            self._emergency_save_step = msg['emergency_save_step']
+
+    def start_step(self):
+        assert not self._is_step_started, "Existing steps must finish before calling start_step()!"
+        self._step_start_time = time.time()
```
Comment (Algy) [inline src/periflow_sdk/__init__.py:129]: `time.time()` is wall clock. How about using `time.monotonic()`?

File: src/periflow_sdk/__init__.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        self._cur_step += 1
+        self._is_step_started = True
+
+    def end_step(self):
+        assert self._is_step_started, "Existing steps must start before calling end_step()!"
+        step_time = time.time() - self._step_start_time
```
Comment (Algy) [inline src/periflow_sdk/__init__.py:135]: `time.time()` is wall clock. How about using `time.monotonic()`?

File: src/periflow_sdk/__init__.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        return self._emergency_save_step == self._cur_step
+
+    def _local_log(self, msg):
+        mode = "a" if self._has_locally_logged else "w"
+        with self._log_path.open(mode=mode) as log_file:
+            log_file.write(f"{json.dumps(msg)}\n")
+        self._has_locally_logged = True
+
+    def log(self, msg: Dict):
+        new_msg = copy.deepcopy(msg)
```
Comment (Algy) [inline src/periflow_sdk/__init__.py:183]: deep copy seems overkill here. You can go for shallow copy here. e.g. `msg.copy()`

File: src/periflow_sdk/__init__.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                    from ipc_connection_failure
+        self._is_step_started = False
+
+    @contextmanager
+    def train_step(self):
+        try:
+            self.start_step()
+            yield
+        finally:
+            self.end_step()
```
Comment (Algy) [inline src/periflow_sdk/__init__.py:171]: Idiom:
```suggestion
        self.start_step()
        try:
            yield
        finally:
            self.end_step()
```

**Review by Algy — COMMENTED** [review]
thx. Here's my PR review.

File: src/periflow_sdk/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DifferentSC) [reply src/periflow_sdk/__init__.py:35]: These variables are for tests

File: src/periflow_sdk/__init__.py (hunk 3 above)
Comment (DifferentSC) [reply src/periflow_sdk/__init__.py:36]: For unit tests

File: src/periflow_sdk/__init__.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+    def is_emergency_save(self):
+        return self._emergency_save_step == self._cur_step
+
+    def _local_log(self, msg):
+        mode = "a" if self._has_locally_logged else "w"
+        with self._log_path.open(mode=mode) as log_file:
```
[the remark this answers is not in the document]
Comment (DifferentSC) [reply src/periflow_sdk/__init__.py:178]: We take this approach because we want to metrics of every iteration is flushed immediately.

**Review by ktaebum — APPROVED** [review]
LGTM
